### PR TITLE
C++ parity, grid_sample functional

### PR DESCRIPTION
**Pull request by lsrock1**
#25883 
I put grid_sample in vision.h with affine grid.

I have a question in string argument(interpolation mode, padding mode)
I reuse torch::native::detail::GridSamplerInterpolation in GridSampler.h instead of using string.
It follows the way that uses reduction enum in loss functions.
I am not sure this is right.

@yf225

### CODE REVIEW

File: caffe2/CMakeLists.txt [hunk 1]
```diff
@@ -573,6 +573,7 @@ if (NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
       ${TORCH_SRC_DIR}/csrc/api/src/nn/options/linear.cpp
       ${TORCH_SRC_DIR}/csrc/api/src/nn/options/normalization.cpp
       ${TORCH_SRC_DIR}/csrc/api/src/nn/options/pooling.cpp
+      ${TORCH_SRC_DIR}/csrc/api/src/nn/options/vision.cpp
```
Comment (yf225) [inline caffe2/CMakeLists.txt]: I think we might not need to create the `torch/csrc/api/src/nn/options/vision.cpp` file, and having the header file should be sufficient :D

File: torch/csrc/api/include/torch/nn/options/vision.h [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+using torch::native::detail::GridSamplerPadding;
+
+namespace nn {
+
+/// Options for a Grid Sample module.
+struct TORCH_API GridSampleOptions {
+  /// interpolation mode to calculate output values. Default: Bilinear
+  TORCH_ARG(torch::GridSamplerInterpolation, mode) = torch::GridSamplerInterpolation::Bilinear;
+  /// padding mode for outside grid values. Default: Zeros
+  TORCH_ARG(torch::GridSamplerPadding, padding_mode) = torch::GridSamplerPadding::Zeros;
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/options/vision.h]: I feel that we can actually use `std::string` here for `mode` and `padding_mode`, to strictly follow the Python version. (And later I will change them to use `c10::variant`-based types as part of a broader change) :D

File: torch/csrc/api/include/torch/nn/options/vision.h [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  /// Specifies perspective to pixel as point. Default: true
+  TORCH_ARG(c10::optional<bool>, align_corners) = c10::nullopt;
+};
+
+} // namespace nn
+} // namespace torch
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/options/vision.h]: nit: we might need a new line here

File: torch/csrc/api/include/torch/nn/functional/vision.h [hunk 4]
```diff
@@ -46,6 +47,21 @@ inline Tensor affine_grid(
...
+    GridSampleOptions options = {}) {
+  
+  if (!options.align_corners().has_value()) {
+    TORCH_WARN("Default grid_sample and affine_grid behavior will be changed ",
+                "to align_corners=False from 1.4.0. ",
+                "See the documentation of grid_sample for details.");
+    options.align_corners(true);
+  }
+  
+  return torch::grid_sampler(input, grid, static_cast<int64_t>(options.mode()), static_cast<int64_t>(options.padding_mode()), options.align_corners().value());
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/vision.h]: It would be awesome to directly translate the Python logic to the C++ version:
https://github.com/pytorch/pytorch/blob/783c9c84450f9b75d8cb9752062e961b7449e2c6/torch/nn/functional.py#L2675-L2702
including the temporary value `mode_enum` and `padding_mode_enum` for the enums. Thanks so much for your help! :D

File: test/cpp/api/functional.cpp [hunk 5]
```diff
@@ -215,6 +215,65 @@ TEST_F(FunctionalTest, HingeEmbeddingLoss) {
...
+  // bilinear, reflection, true
+  options = GridSampleOptions()
+                  .mode(torch::GridSamplerInterpolation::Bilinear)
+                  .padding_mode(torch::GridSamplerPadding::Reflection)
+                  .align_corners(true);
+  output = F::grid_sample(input, grid, options);
+  expected = torch::tensor({{{{1., 0., 1.}, {3., 4., 5.}, {7., 8., 7.}}}}, torch::kFloat);
+
+  ASSERT_TRUE(output.allclose(expected));
+}
```
Comment (yf225) [inline test/cpp/api/functional.cpp]: Thanks so much for the thorough tests! I think we could add another one for `F::grid_sample(input, grid)` (i.e. "no-options") use case. Thanks a lot! :D

**Review by yf225 — COMMENTED** [review]
@lsrock1 Thanks so much for the awesome work! I left some comments :D

**Comment by lsrock1** [comment]
@yf225 updated!

File: setup.py [hunk 6]
```diff
@@ -800,6 +800,7 @@ def print_box(msg):
                 'include/ATen/cuda/detail/*.h',
                 'include/ATen/cudnn/*.h',
                 'include/ATen/detail/*.h',
+                'include/Aten/native/*.h',
```
Comment (yf225) [inline setup.py]: I believe we can remove this as well :D

File: torch/csrc/api/src/nn/options/vision.cpp [hunk 7]
```diff
@@ -0,0 +1 @@
+#include <torch/nn/options/vision.h>
```
Comment (yf225) [inline torch/csrc/api/src/nn/options/vision.cpp]: I believe we can remove this `torch/csrc/api/src/nn/options/vision.cpp` file :D

File: tools/build_variables.py [hunk 8]
```diff
@@ -229,6 +229,7 @@ def add_torch_libs():
         "torch/csrc/api/src/nn/options/linear.cpp",
         "torch/csrc/api/src/nn/options/normalization.cpp",
         "torch/csrc/api/src/nn/options/pooling.cpp",
+        "torch/csrc/api/src/nn/options/vision.cpp",
```
Comment (yf225) [inline tools/build_variables.py]: I believe we can remove this `torch/csrc/api/src/nn/options/vision.cpp` file :D

File: torch/csrc/api/include/torch/nn/functional/vision.h [hunk 9]
```diff
@@ -46,6 +47,53 @@ inline Tensor affine_grid(
...
+
+  if ((options.mode().compare("bilinear") != 0) && (options.mode().compare("nearest") != 0)) {
+    TORCH_CHECK(false, "nn::functional::grid_sample(): expected mode to be ",
+                         "'bilinear' or 'nearest', but got: '", options.mode(), "'");
+  }
+
+  if ((options.padding_mode().compare("zeros") != 0) && 
+      (options.padding_mode().compare("border") != 0) && 
+      (options.padding_mode().compare("reflection") != 0)) {
+    TORCH_CHECK(false, "nn.functional.grid_sample(): expected padding_mode ",
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/vision.h]: nit:
```suggestion
    TORCH_CHECK(false, "nn::functional::grid_sample(): expected padding_mode ",
```

File: torch/csrc/api/include/torch/nn/functional/vision.h [hunk 10]
```diff
@@ -46,6 +47,53 @@ inline Tensor affine_grid(
...
+  }
+  else { /// padding_mode == 'reflection'
+    padding_mode_enum = 2;
+  }
+  
+  if (!options.align_corners().has_value()) {
+    TORCH_WARN("Default grid_sample and affine_grid behavior will be changed ",
+                "to align_corners=False from 1.4.0. ",
+                "See the documentation of grid_sample for details.");
+    options.align_corners(true);
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/vision.h]: Curious is there a reason why the warning message and the `options.align_corners(true)` setting different from the Python version?

File: torch/csrc/api/include/torch/nn/options/vision.h [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+namespace torch {
+namespace nn {
+
+/// Options for a Grid Sample module.
+struct TORCH_API GridSampleOptions {
+  /// interpolation mode to calculate output values. Default: Bilinear
+  TORCH_ARG(std::string, mode) = "bilinear";
+  /// padding mode for outside grid values. Default: Zeros
+  TORCH_ARG(std::string, padding_mode) = "zeros";
+  /// Specifies perspective to pixel as point. Default: true
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/options/vision.h]: I believe we can change it to match the Python version:
```suggestion
  /// Specifies perspective to pixel as point. Default: false
```

File: torch/csrc/api/include/torch/nn/options/vision.h [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+  TORCH_ARG(c10::optional<bool>, align_corners) = c10::nullopt;
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/options/vision.h]: I believe we can change it to
```suggestion
  TORCH_ARG(bool, align_corners) = false;
```
to simplify the argument type and also match the Python version :D

**Review by yf225 — COMMENTED** [review]
@lsrock1 Thanks a lot for the update! I left some minor comments.

File: torch/csrc/api/include/torch/nn/functional/vision.h (hunk 10 above)
Comment (lsrock1) [reply torch/csrc/api/include/torch/nn/functional/vision.h]: Sry, i saw the old 1.3 doc
https://pytorch.org/docs/stable/_modules/torch/nn/functional.html#grid_sample
i will fix it!

**Comment by lsrock1** [comment]
@yf225 check please!

**Review by yf225 — APPROVED** [review]
@lsrock1 Thanks so much for the awesome work! My apologies that I just realized we do need `torch/csrc/api/src/nn/options/vision.cpp` for Windows builds, and I added it back :D
